magma/backend/coreir_.py

```diff
@@ -81,445 +32,55 @@ def get_instance(self):
         return self.__instance
 
     def reset_instance(self):
-        self.__instance = coreir.Context()
+        self.__instance = Context()
 
     def __init__(self):
-        self.__instance = coreir.Context()
+        self.__instance = Context()
+
+
 CoreIRContextSingleton()
 
+_context_to_modules = {}
+
+
 class CoreIRBackend:
-    context_to_modules_map = {}
-    def __init__(self, context=None, check_context_is_default=True):
+    def __init__(self, context=None):
+        singleton = CoreIRContextSingleton().get_instance()
         if context is None:
-            context = CoreIRContextSingleton().get_instance()
-        elif check_context_is_default & (context != CoreIRContextSingleton().get_instance()):
-            warn("Creating CoreIRBackend with non-singleton CoreIR context. "
-                 "If you're sure you want to do this, set check_context_is_default "
-                 "when initializing the CoreIRBackend.")
-        if context not in CoreIRBackend.context_to_modules_map:
-            CoreIRBackend.context_to_modules_map[context] = {}
-        self.modules = CoreIRBackend.context_to_modules_map[context]
+            context = singleton
+        if context is not singleton:
+            logger.warning("Creating CoreIRBackend with non-singleton CoreIR "
+                           "context.")
+        self.modules = _context_to_modules.setdefault(context, {})
         self.context = context
         self.libs = keydefaultdict(self.context.get_lib)
         self.libs_used = set()
-        self.__constant_cache = {}
-        self.__unique_concat_id = -1
-
-    def check_interface(self, definition):
-        # for now only allow Bit, Array, or Tuple
-        def check_type(port, errorMessage=""):
-            if isinstance(port, ArrayKind):
-                check_type(port.T, errorMessage.format("Array({}, {})").format(
-                    str(port.N), "{}"))
-            elif isinstance(port, TupleKind):
-                for (k, t) in zip(port.Ks, port.Ts):
-                    check_type(t, errorMessage.format("Tuple({}:{})".format(k, "{}")))
-            elif isinstance(port, (BitKind, ClockKind, EnableKind, ResetKind, AsyncResetKind, ResetNKind, AsyncResetNKind)):
-                return
-            else:
-                raise CoreIRBackendError(errorMessage.format(str(port)))
-        for name, port in definition.interface.ports.items():
-            check_type(type(port), 'Error: Argument {} must be comprised only of Bit, Array, or Tuple')
-
-    def get_type(self, port):
```

Comment:
I also recall @David-Durst 's code using this method (and possible others) from the corer backend directly, so either we should try to preserve the old interface so we don't break his code, or work with him on a plan to update it.

Reply:
Actually, reading through these changes, perhaps the usage I remember was restricted to the coreir frontend? (Which has been updated, so should be okay) @David-Durst can you comment on whether these changes to the coreir backend will affect your code?

Reply:
I need to port renaming preserved (converting between CoreIR's .I ports and magma's .in names) Otherwise, I think the I'm mainly using https://github.com/phanrahan/magma/blob/master/magma/frontend/coreir_.py directly. Unfortunately, CI won't be useful in debugging these issues since I'm still on a CoreIR branch.

Reply:
@David-Durst no functionality has changed. Most likely the same functionality is now available to in `frontend/coreir_.py`. Otherwise, it may have moved to `backend/coreir_utils.py`, in which case I can expose it for you and you'll just have to modify the call-sites marginally.

Reply:
Ok, then this should work for me. I'll make changes to my code if it breaks. lgtm.